cores/esp8266/umm_malloc/umm_malloc.cpp

```diff
@@ -519,8 +540,92 @@ extern int umm_last_fail_alloc_size;
+  storage for current state. Expanded UMM_CRITICAL_... to  use unique
+  identifiers. This helpt facilitate gather function specific  timing
+  information.
+
+  Replace printf with something that is ROM or IRAM based so that a printf
+  that occurs during an ISR malloc/new does not cause a crash. To avoid any
+  reentry issue it should also avoid doing malloc lib calls.
+
+  Refactor realloc to avoid memcpy/memmove while in critical section. This is
+  only effective when realloc is called with interrupts enabled. The copy
+  process alone can take over 10us (when copying more than ~498 bytes with a
+  80MHz CPU clock). It would be good practice for an ISR to avoid realloc.
+  Note, while doing this might initially sound scary, this appears to be very
+  stable. It ran on my troublesome sketch for over 3 weeks until I got back from
+  vacation and  flashed an update. Troublesome sketch - runs ESPAsyncTCP, with
+  modified fauxmo emulation for 10 devices. It receives lost of Network traffic
+  related to uPnP scans, which includes lots of TCP connects disconnects RSTs
+  related to uPnP discovery.
+
+  I have clocked umm_info critical lock time taking as much as 180us. A common
+  use for the umm_info call is to get the free heap result. It is common
+  to try and closely monitor free heap as a method to detect memory leaks.
+  This may result in frequent calls to umm_info. There has not been a clear
+  test case that shows an issue yet; however, I and others think they are or
+  have had crashes related to this.
+
+  I have added code that adjusts the running free heap number from _umm_malloc,
+  _umm_realloc, and _umm_free. Removing the need to do a long interrupts
+  disabled calculation via _umm_info.
+
+  Build optional, min/max time measurements for locks held while in info,
+  malloc, realloc, and free. Also, maintain a count of how many times each is
+  called with INTLEVEL set.
+
+ */
+
+#if defined(DEBUG_ESP_PORT) || defined(DEBUG_ESP_ISR)
+#define DEBUG_IRAM_ATTR ICACHE_RAM_ATTR
+/*
+   Printing from the malloc routines is tricky. Since a lot of library calls
+  will want to do malloc.
+
+  Objective:  To be able to print "last gasp" diagnostic messages
+  when interrupts are disabled and w/o availability of heap resources.
+*/
+
+// ROM _putc1, ignores CRs and sends CR/LF for LF, newline.
+// Always returns character sent.
+int constexpr (*_rom_putc1)(int) = (int (*)(int))0x40001dcc;
+void uart_buff_switch(uint8_t);
+
+int DEBUG_IRAM_ATTR _isr_safe_printf_P(const char *fmt, ...) __attribute__((format(printf, 1, 2)));
+int DEBUG_IRAM_ATTR _isr_safe_printf_P(const char *fmt, ...) {
+#ifdef DEBUG_ESP_PORT
+#define VALUE(x) __STRINGIFY(x)
+  // Preprocessor and compiler together will optimize away the if.
+  if (strcmp("Serial1", VALUE(DEBUG_ESP_PORT)) == 0) {
+    uart_buff_switch(1U);
+  } else {
+    uart_buff_switch(0U);
+  }
+#else
+  uart_buff_switch(0U); // Side effect, clears RX FIFO
+#endif
+  size_t str_len = ets_strlen(fmt);
```

Comment:
The format strings live in progmem, so ets_strlen() is going to bomb with an access error, I believe.

Reply:
It has been working for me. Before using ets_strlen I checked one of the assembly listings on the net for the ROM. It starts by verifying the pointer alignment. If it is aligned by 4, it drops down to the fast path of reading a word at a time, masking a byte out at a time and checking for 0. So unless I misread we should be safe as long as we start with a PROGMEM string that is aligned by 4. I just rechecked `tools/sdk/libc/xtensa-lx106-elf/include/sys/pgmspace.h` 
```
#define PSTR(s)            (__extension__({static const char __c[] __attribute__((__aligned__(4))) PROGMEM = (s); &__c[0];}))
```
I think this is relatively new. Let me know if I have errored in my thinking.

Reply:
You're correct there.  I added that alignment in 2.5.0 (IIRC) because I rewrote the _P routines in the std. lib to be much faster w/aligned data.  I didn't know ets_strlen had that optimization.  I guess that's a lucky coincidence on our part!

Reply:
Could you add a note to that effect, just for our editication?  There may be other, more memory constrained branches somewhere looking at our code w/o that alignment.  It'll crash for them, but at least they'd get a note in the source why it worked for is?

Reply:
After further thought, I am thinking this comment before ets_strlen would be good:
```
  /* 
    To use ets_strlen() and ets_memcpy() safely with PROGMEM, flash storage,
    the PROGMEM address must be word (4 bytes) aligned. The destination
    address for ets_memcpy must also be word-aligned. We also round the
    buf_len up to the nearest word boundary. So that all transfers will be
    whole words. 
  */
```